### ISSUE TITLE
Compaction error: IO error: .../xxxxx.ldb: Too many open files

### ISSUE THREAD

**Issue by rchunping** [google/leveldb#401]
I also read the  [issue 181](https://github.com/google/leveldb/issues/181#issuecomment-55006885)

> LevelDB's above a certain size (about 40 GB) seems to cause leveldb to open every single file in the database without closing anything in between.
> 
> Also, it seems it opens every file twice, for some reason.

My problem is almost the same.

OS: FreeBSD 10.1-RELEASE amd64
Leveldb: master branch, ( also test  1.18,1.17,...1.14 )
Dataset:  99G with snappy compressed, 58612 *.sst files.
ulimit -n: 706995
kern.maxfiles: 785557
kern.maxfilesperproc: 706995

The dataset was generated by leveldb 1.8.0 , running several months.
Last week , I restart the server , then the issue occurred.

It seems open every *.sst file twice,and not close them.

``` sh
$ fstat -m|grep leveldb|wc
  117223 1055007 8668825
```

58612 \* 2 ~= 117223  <  **706995**  (system limit)

``` sh
$ fstat -m|grep leveldb
USER     CMD          PID   FD MOUNT      INUM MODE         SZ|DV R/W
root     leveldb-tools 67098   67 /         92326 -rw-r--r--  1594319  r
root     leveldb-tools 67098   68 /         92326 -rw-r--r--  1594319  r
root     leveldb-tools 67098   69 /         45578 -rw-r--r--  2124846  r
root     leveldb-tools 67098   70 /         45578 -rw-r--r--  2124846  r
root     leveldb-tools 67098   71 /         45579 -rw-r--r--  2123789  r
root     leveldb-tools 67098   72 /         45579 -rw-r--r--  2123789  r
root     leveldb-tools 67098   73 /         45580 -rw-r--r--  2125455  r
root     leveldb-tools 67098   74 /         45580 -rw-r--r--  2125455  r
root     leveldb-tools 67098   75 /         45581 -rw-r--r--  2123795  r
root     leveldb-tools 67098   76 /         45581 -rw-r--r--  2123795  r
root     leveldb-tools 67098   77 /         45582 -rw-r--r--  2122645  r
root     leveldb-tools 67098   78 /         45582 -rw-r--r--  2122645  r
root     leveldb-tools 67098   79 /         45583 -rw-r--r--  2119487  r
root     leveldb-tools 67098   80 /         45583 -rw-r--r--  2119487  r
root     leveldb-tools 67098   81 /         45584 -rw-r--r--  2117737  r
root     leveldb-tools 67098   82 /         45584 -rw-r--r--  2117737  r
... more ....
```

as above , each file open twice (the same inode num: 92326,92326,45578,45578,...)

``` sh
$ tail -f LOG
2016/08/10-11:17:48.121149 802006400 Recovering log #18223888
2016/08/10-11:17:48.329778 802006400 Delete type=2 #18223889
2016/08/10-11:17:48.333491 802006400 Delete type=3 #18223887
2016/08/10-11:17:48.333993 802006400 Delete type=0 #18223888
2016/08/10-11:17:48.388989 802007400 Compacting 58608@0 + 0@1 files
2016/08/10-11:20:14.324576 802007400 compacted to: files[ 58608 0 0 0 0 0 0 ]
2016/08/10-11:20:14.325108 802007400 Compaction error: IO error: ..../leveldb/18223891.ldb: Too many open files
```

After the IO error, the open files reduse to 87580

``` sh
fstat -m | grep leveldb | wc
   87580  788220 6476498
```

And the program cost 100% CPU

``` sh
  PID USERNAME    THR PRI NICE   SIZE    RES STATE   C   TIME    WCPU COMMAND
67098 root          6  35    0  5293M  3607M uwait   4  48:55 100.00% leveldb-tools
```

But there is no disk io at all

``` sh
  PID USERNAME     VCSW  IVCSW   READ  WRITE  FAULT  TOTAL PERCENT COMMAND
67098 root            0    216      0      0      0      0   0.00% leveldb-tools
```

Then ,can't seek, can't get, can't put .....

I'v tried change leveldb_options_set_max_open_files()  , 100, 1024, 400000, but it not worked.

**Comment by cmumford** [comment]
v1.8 is fairly old - [v1.19](https://github.com/google/leveldb/releases/tag/v1.19) was just released. There have been at least two changes related to max open files addressed since v1.8. However, your output table has a "*.ldb" extension - which is part of v1.14. I'm guessing you really meant v1.18? Can you try with v1.19, which has 06a191b8, which _might_ fix your problem?

**Comment by rchunping** [comment]
@cmumford  thanks for your reply.

Actualy, the dataset was generated by leveldb 1.8(very old). and last week , the server's disk error cause some .sst files lost.

I copy the dataset to another server, and wrote a new tool try to repair the data.

The new tool is based on leveldb master(2016-08-08) + snappy 1.1.3.

after leveldb_repair_db(), the LOG : 
```
...
2016/08/12-00:51:35.882727 802006400 Table #17120980: 3861 entries OK
2016/08/12-00:51:35.989462 802006400 Table #17354499: 3572 entries OK
2016/08/12-00:51:36.092376 802006400 Archiving /leveldb/MANIFEST-18223896: OK
2016/08/12-00:51:36.155131 802006400 **** Repaired leveldb /leveldb; recovered 58624 files; 120998381410 bytes. Some data may have been lost. ****
```

then , leveldb_open() , the LOG : 
```
...
2016/08/12-00:58:55.777293 802006400 Recovering log #18223897
2016/08/12-00:58:56.002376 802006400 Delete type=3 #18223898
2016/08/12-00:58:56.017200 802006400 Delete type=0 #18223897
2016/08/12-00:58:56.082653 802007400 Compacting 58606@0 + 0@1 files
2016/08/12-01:05:45.973430 802007400 compacted to: files[ 58606 0 0 0 0 0 0 ]
2016/08/12-01:05:45.973909 802007400 Compaction error: IO error: /leveldb/18223901.ldb: Too many open files
```

**Comment by cmumford** [comment]
Thx @rchunping. Can you give it a try with v1.19 (with 06a191b), just on the off chance it makes a difference?

**Comment by rchunping** [comment]
@cmumford Thansk.

I'v tried the 06a191b
```
3080a45 * HEAD -> master v1.19 origin/master origin/HEAD Increase leveldb version to 1.19.                                                                                                                          cmumford       1 day
fa6dc01 * A zippy change broke test assumptions about the size of compressed output. Fix the tests by allowing more slop in zippy's behavior. ------------- Created by MOE: https://github.com/google/moe MOE_MIGRA$sanjay         3 months
06a191b * fix problems in LevelDB's caching code                                                                                                                                                                    m3b            3 months
a7bff69 * Fix LevelDB build when asserts are enabled in release builds. (#367)                                                                                                                                      John Abd-E…   4 months
....
```

there's compile error:
```
$ env CC=g++47 CXX=g++47 CPP=cpp47 gmake
....
db/c_test.c: In function 'char* FilterCreate(void*, const char* const*, const size_t*, int, size_t*)':
db/c_test.c:144:26: error: invalid conversion from 'void*' to 'char*' [-fpermissive]
gmake: *** [Makefile:408: out-static/db/c_test.o] Error 1
```

use this little patch to fix the error:
```patch
diff --git a/db/c_test.c b/db/c_test.c
index 7cd5ee0..d70c9e9 100644
--- a/db/c_test.c
+++ b/db/c_test.c
@@ -141,7 +141,7 @@ static char* FilterCreate(
     int num_keys,
     size_t* filter_length) {
   *filter_length = 4;
-  char* result = malloc(4);
+  char* result =(char *) malloc(4);
   memcpy(result, "fake", 4);
   return result;
 }
```

leveldb_repair_db() : 
```
...
2016/08/12-14:02:07.929734 802006400 Table #17120980: 3861 entries OK
2016/08/12-14:02:07.977840 802006400 Table #17354499: 3572 entries OK
2016/08/12-14:02:08.098454 802006400 Archiving /leveldb/MANIFEST-18223901: OK
2016/08/12-14:02:08.187612 802006400 **** Repaired leveldb /leveldb; recovered 58618 files; 120986359200 bytes. Some data may have been lost. ****
```
leveldb_open() : 
```
2016/08/12-16:10:23.650254 802006400 Delete type=2 #18223896
2016/08/12-16:10:23.653825 802006400 Delete type=3 #1
2016/08/12-16:10:23.745875 802007400 Compacting 58605@0 + 0@1 files
2016/08/12-16:16:14.683455 802007400 compacted to: files[ 58605 0 0 0 0 0 0 ]
2016/08/12-16:16:14.683980 802007400 Compaction error: IO error: /leveldb/18223899.ldb: Too many open files
```

**Comment by rchunping** [comment]
@cmumford  Update:

After doing some tests, I think it's not leveldb's issue. sorry.

I'm not good at c++, so I use [goleveldb](https://github.com/syndtr/goleveldb) (leveldb's golang implementation) for test.

#### 1.  first , I tried leveldb.RecoverFile() 

there are some error info in LOG
```
...
leveldb/table: Writer: keys are not in increasing order: "\xff\x01item/98830906\x01\x01\xfe\x88IZ\x00\x00\x00", "\xff\x01item/98830906\x01\x01?        
...
leveldb/table: corruption on table-footer (pos=14697): bad magic number [file=18223895.ldb]       
...                                                   
```
looks like some table files were corrupted.

#### 2. modify goleveldb's recover code to force rebuild each table file , and drop the corrupted files.

and still failed, after RecoverFile(), it report "file missing"


#### 3. the .ldb files which rebuilded in step(2) should be ok. so I continue modify goleveldb's code , and dump key/value pairs from each .ldb file.

after that, I got a 290G dump file. (about 216385899 key/value pairs)

#### 4. import the dump file to a new database.

this step took about 6 hours to completed.
why 6 hours?
after import about 60G key/value data, leveldb report "Too many L0 files; waiting....", and block write about 3 hours.

#### Finally, open the new database use google/leveldb , it's ok now. 

most of the data was come back. and because of some corrupted .ldb files, a few thousands of k/v pair were lost.

Now, the database is runing on a new server with raid1 storage.

**Comment by cmumford** [comment]
goleveldb doesn't just provide a go wrapper to leveldb - it is a completely different project written entirely in go.

**Comment by vnsanilkumar** [comment]
@Cmumford, why you are closing the issue by knowing it(what said by @rchunping) is not the actual fix. I am also facing the same problem after running repairdb on that database file. Please let me know, how we can solve it without going to use other binding. Thanks for your help in advance.

**Comment by rchunping** [comment]
@vnsanilkumar how many .ldb files in your database directory?

**Comment by vnsanilkumar** [comment]
i am running 12 db instances in parallel. total ldb files are 27958 in all those folders(db folder).

**Comment by rchunping** [comment]
@vnsanilkumar after repairdb,all of the .ldb files are in level-0, i guess this is one of the problems.

If you want to dump you data from that db and import it to another new db, i can push my tool to github this weekend.

But there are some shortcomings:
1. Not 100% of your data will come back,but at least most of them.
2. If you update some key frequently, there may be multiple values  under  the same key in Level-0 tabel file. I try to use the latest value.
3. because of (2), some deleted key may be  "come back".

**Comment by vnsanilkumar** [comment]
@rchunping thanks for the information. Can i solve this one by increasing the ulimit of system ?

**Comment by rchunping** [comment]
@vnsanilkumar you can try it first.

but in my issue, number of opend files by leveldb is less than my system's limit   (58612 * 2 ~= 117223 < 706995 (system limit))

**Comment by vnsanilkumar** [comment]
@rchunping  thanks for support. I am unable to solve this one by increasing ulimit values as you said. I am waiting for your tool. I will try with latest version as suggested by @cmumford . Through me any thoughts in between.

**Comment by rchunping** [comment]
@vnsanilkumar my tool : https://github.com/rchunping/leveldb-tools

before you start, please **DO BACKUP YOUR DATA** first.

try actions in this order:

```
1. export 
2. if not work,  repair and then export
3. if still not work, export2
4. import to new db
```

**Comment by cmumford** [comment]
@vnsanilkumar I think I may have misunderstood the comment made by @rchunping. I thought he was saying that the original bug was encountered under goleveldb, but after a second reading he may have just tried to repair using goleveldb. @rchunping can you clarify?

**Comment by rchunping** [comment]
@cmumford the bug was encountered under google/leveldb after repairdb().

Because I am not good at C++, I write a tool [leveldb-tools](https://github.com/rchunping/leveldb-tools) (use  syndtr/goleveldb) to dump key/value data from each .ldb file.

**Comment by cmumford** [comment]
This should not have been closed - and my apologies for waiting so long before reopening.

**Comment by cmumford** [comment]
This should now be fixed in 646c3588de84 *if* the POSIX Env is being used.

Some background: leveldb has an max_open_files option, but that is ignored by RandomAccessFile. An unbounded number of RandomAccessFile instances will be created during compaction - hence this bug. On 64-bit systems the first 1000 RandomAccessFile instances are memory mapped which does not consume a file descriptor. Beyond that fd's will be used (for popen). On 32-bit systems only fd's are used.

This fix doesn't strictly limit fd usage to a hard value, but it should be extremely unlikely that any limit is reached (assuming the hosting application doesn't open a huge number of files while also using leveldb).
